src/vm/slot.rs

```diff
@@ -11,44 +11,91 @@ use atomic::Atomic;
+/// non-reference values**, such as `null`, `true`, etc.  When the slot is holding an object
+/// reference, we can load the object reference from it, and we can update the object reference in
+/// it after the GC moves the object.
+///
+/// # The `Slot` trait has pointer semantics
+///
+/// A `Slot` value *points to* a slot, and is not the slot itself.  In fact, the simplest
+/// implementation of the `Slot` trait ([`SimpleSlot`], see below) can simply contain the address of
+/// the slot.
+///
+/// A `Slot` can be [copied](std::marker::Copy), and the copied `Slot` instance points to the same
+/// slot.
+///
+/// # How to implement `Slot`?
+///
+/// If a reference field of a VM is just a word that holds the pointer to an object, and uses the 0
+/// word as the null pointer, it can use the default [`SimpleSlot`] we provide.  It simply contains
+/// a pointer to a memory location that holds an address.
+///
+/// ```rust
+/// pub struct SimpleSlot {
+///     slot_addr: *mut Atomic<Address>,
+/// }
+/// ```
+///
+/// In other cases, the VM need to implement its own `Slot` instances.
 ///
 /// For example:
-/// -   The VM uses compressed pointer (Compressed OOP in OpenJDK's terminology), where the heap
-///     size is limited, and a 64-bit pointer is stored in a 32-bit slot.
-/// -   The VM uses tagged pointer, where some bits of a word are used as metadata while the rest
-///     are used as pointer.
-/// -   A field holds a pointer to the middle of an object (an object field, or an array element,
-///     or some arbitrary offset) for some reasons.
+/// -   The VM uses **compressed pointers** (Compressed OOPs in OpenJDK's terminology), where the
+///     heap size is limited, and a 64-bit pointer is stored in a 32-bit slot.
+/// -   The VM uses **tagged pointers**, where some bits of a word are used as metadata while the
+///     rest are used as pointer.
+/// -   The VM uses **offsetted pointers**, i.e. the value of the field is an address at an offset
+///     from the [`ObjectReference`] of the target object.  Such offsetted pointers are usually used
+///     to represent **interior pointers**, i.e. pointers to an object field, an array element, etc.
+///
+/// If needed, the implementation of `Slot` can contain not only the pointer, but also additional
+/// information. The `OffsetSlot` example below also contains an offset which can be used when
+/// decoding the pointer. See `src/vm/tests/mock_tests/mock_test_slots.rs` for more concrete
+/// examples.
 ///
-/// When loading, `Slot::load` shall decode its internal representation to a "regular"
-/// `ObjectReference`.  The implementation can do this with any appropriate operations, usually
-/// shifting and masking bits or subtracting offset from the address.  By doing this conversion,
-/// MMTk can implement GC algorithms in a VM-neutral way, knowing only `ObjectReference`.
+/// ```rust
+/// pub struct OffsetSlot {
```

Comment:
Note that a `Slot` like this will not be word-sized so may not have the best performance. It should be flagged.